docs/env-vars.md
# ENVIRONMENT VARIABLES

| Environment Variable                     | Description                                                                         | Default Value                           |
|------------------------------------------|-------------------------------------------------------------------------------------|-----------------------------------------|
| `AUTHENTICATION_DISABLED`                | Disable authentication                                                              | `false`                                 |
| `AUTHENTICATOR_OIDC_CLIENT_IDS`          | Set allowed client IDs (comma separated)                                            |                                         |
| `AUTHENTICATOR_OIDC_ISSUER_URL`          | Issuer URL of the clients                                                           |                                         |
| `AUTHENTICATOR_OIDC_REQUIRED_AUDIENCE`   | Enforce an "audience" to be present in the access token                             |                                         |
| `AUTHENTICATOR_OIDC_TLS_CA_CERTIFICATES` | Enable additional TLS certificates for communication with the SSO server            |                                         |
| `AUTHENTICATOR_OIDC_TLS_INSECURE`        | Allow insecure TLS connections with the SSO server                                  |                                         |
| `AUTH_CONFIGURATION`                     | Location of the AuthNZ configuration file                                           |                                         |
| `AUTH_DISABLED`                          | Disable authentication and authorization                                            | `false`                                 |
| `CLIENT_TLS_CA_CERTIFICATES`             | Additional certificates which will be added as trust anchors                        |                                         |
| `CLIENT_TLS_INSECURE`                    | Make the TLS client insecure, disabling all validation                              | `false`                                 |
| `HTTP_SERVER_BIND_ADDR`                  | Address to listen on                                                                | `::1`                                   |
| `HTTP_SERVER_DISABLE_LOG`                | Disable the request log                                                             | `false`                                 |
| `HTTP_SERVER_JSON_LIMIT`                 | JSON request limit                                                                  | `2 MiB`                                 |
| `HTTP_SERVER_REQUEST_LIMIT`              | Overall request limit                                                               | `256 KiB`                               |
| `HTTP_SERVER_TLS_CERTIFICATE_FILE`       | Path to the TLS certificate in PEM format                                           |                                         |
| `HTTP_SERVER_TLS_ENABLED`                | Enable TLS                                                                          | `false`                                 |
| `HTTP_SERVER_TLS_KEY_FILE`               | Path to the TLS key file in PEM format                                              |                                         |
| `HTTP_SERVER_WORKERS`                    | Number of worker threads, defaults to zero, which falls back to the number of cores | `0`                                     |
| `IMPORTER_CONCURRENCY`                   | The maximum number of jobs run simultaneously by the importer                       | `1`                                     |
| `IMPORTER_WORKING_DIR`                   | Where the importer downloads documents prior to ingesting them                      | `tempdir`                               |
| `OIDC_PROVIDER_CLIENT_ID`                | OIDC client ID used for retrieving access tokens                                    |                                         |
| `OIDC_PROVIDER_CLIENT_SECRET`            | Secret matching the OIDC client ID                                                  |                                         |
| `OIDC_PROVIDER_ISSUER_URL`               | OIDC issuer to request access tokens from                                           |                                         |
| `OIDC_PROVIDER_REFRESH_BEFORE`           | Duration an access token must still be valid before requesting a new one            | `30s`                                   |
| `OIDC_PROVIDER_TLS_INSECURE`             | Insecure TLS when contacting the OIDC issuer                                        | `false`                                 |
| `OPENAI_API_KEY`                         | OpenAI access key                                                                   |                                         |
| `OPENAI_API_BASE`                        | To set the base URL path for API requests                                           | `https://api.openapi.com/v1`            |
| `OPENAI_MODEL`                           | OpenAI model                                                                        | `gpt-4o`                                |
| `TRUSTD_DB_HOST`                         | Database address                                                                    | `localhost`                             |
| `TRUSTD_DB_MAX_CONN`                     | Database max connections                                                            | `75`                                    |
| `TRUSTD_DB_MIN_CONN`                     | Database min connections                                                            | `25`                                    |
| `TRUSTD_DB_CONNECT_TIMEOUT`              | Database connect timeout (humantime)                                                | `8s`                                    |
| `TRUSTD_DB_ACQUIRE_TIMEOUT`              | Database acquire timeout (humantime)                                                | `8s`                                    |
| `TRUSTD_DB_MAX_LIFETIME`                 | Database max lifetime (humantime)                                                   | `7200s`                                 |
| `TRUSTD_DB_IDLE_TIMEOUT`                 | Database idle timeout (humantime)                                                   | `600s`                                  |
| `TRUSTD_DB_NAME`                         | Database name                                                                       | `trustify`                              |
| `TRUSTD_DB_PASSWORD`                     | Database password                                                                   | `trustify`                              |
| `TRUSTD_DB_PORT`                         | Database port                                                                       | `5432`                                  |
| `TRUSTD_DB_USER`                         | Database username                                                                   | `postgres`                              |
| `TRUSTD_ISSUER_URL`                      | Issuer URL for `--devmode`                                                          | `http://localhost:8090/realms/trustify` |
| `TRUSTD_MAX_CACHE_SIZE`                  | Maximum size of the graph cache.                                                    | `200 MiB`                               |
| `TRUSTD_S3_ACCESS_KEY`                   | S3 access key                                                                       |                                         |
| `TRUSTD_S3_BUCKET`                       | S3 bucket name                                                                      |                                         |
| `TRUSTD_S3_REGION`                       | S3 region name                                                                      |                                         |
| `TRUSTD_S3_SECRET_KEY`                   | S3 secret key                                                                       |                                         |
| `TRUSTD_STORAGE_FS_PATH`                 | Path for storage file system strategy                                               | `./.trustify/storage`                   |
| `TRUSTD_STORAGE_STRATEGY`                | Specifies the storage strategy to use                                               | `File system`                           |
| `TRUSTD_WITH_GRAPHQL`                    | Allows enabling the GraphQL endpoint                                                | `false`                                 |
| `UI_CLIENT_ID`                           | Client ID used by the UI                                                            | `frontend`                              |
| `UI_ISSUER_URL`                          | Issuer URL used by the UI                                                           | `http://localhost:8090/realms/trustify` |
| `UI_SCOPE`                               | Scopes to request                                                                   | `openid`                                |

## Development 

| Environment Variable         | Description                                                                                                 | Default Value |
|------------------------------|-------------------------------------------------------------------------------------------------------------|---------------|
| `EXTERNAL_TEST_DB`           | Run tests against external test database if set                                                             |               |
| `EXTERNAL_TEST_DB_BOOTSTRAP` | Run tests against external test database if set                                                             |               |
| `MEM_LIMIT_MB`               | Set memory limit for tests that use TrustifyContext, shows the memory usage when the test reaches the limit | `500 MiB`     |


entity/CHANGES.md
After some tests, here's an update:

```mermaid
erDiagram
  Sbom {
    int sbom_id PK "inherited"
    string id "inherited"
    string name "inherited"
    string namespace "inherited"

    uuid sha256
  }

  Package {
    int sbom_uid FK "inherited"
    string id "inherited"
    string name "inherited"
  }

  Node {
    int sbom_id
    string id
  }

  Edge {
    int sbom_id

    string left_id
    option~string~ left_namespace

    enum~relationship~ type

    string right_id
    option~string~ right_namespace
  }

  Sbom 1 to 0+ Package : contains

  BasePurl {
    uuid id
    string type
    string name
    option~string~ namespace
  }
  VersionedPurl {
    string version
  }
  QualifiedPurl {
    jsonb qualifiers
  }

  QualifiedPurl 0+ to 1 VersionedPurl: refines
  VersionedPurl 0+ to 1 BasePurl: refines

  CPE

  Package 1 optionally to 0+ QualifiedPurl: references
  Package 1 optionally to 0+ CPE: references

  Node 1 to zero or one Package: inherits
  Node 1 to zero or one Sbom: inherits

  Edge 1 to zero or one Node: left
  Edge 1 to zero or one Node: right
```

Most notable changes:

* We keep an "SBOM ID" which we assign to ensure we can use multiple versions of SBOMs using the same namespace
* Drop the "describes" tables, as those come with the "describes" relationship
* Create a common `sbom_node` table, having `sbom` and `sbom_package` inherit from those
* Have the `package_relates_to_package` reference entries in `sbom_node`

Unresolved issues:

* A relationship can point outside the current SBOM. That's indicated by the `left_namespace` and `right_namespace`:
  * It might be that both ends are external to this document. I don't think we should support this, even if the spec might.
  * How do we handle the case where the relationship points to a target for which we have multiple SBOMs?


common/src/config.rs
use anyhow::anyhow;
use clap::ValueEnum;
use humantime::parse_duration;
use std::env;
use time::ext::NumericalStdDuration;

const DB_NAME: &str = "trustify";
const DB_USER: &str = "postgres";
const DB_PASS: &str = "trustify";
const DB_HOST: &str = "localhost";
const DB_PORT: u16 = 5432;
const DB_MAX_CONN: u32 = 75;
const DB_MIN_CONN: u32 = 25;
const DB_CONNECT_TIMEOUT: u64 = 8;
const DB_ACQUIRE_TIMEOUT: u64 = 8;
const DB_MAX_LIFETIME: u64 = 7200;
const DB_IDLE_TIMEOUT: u64 = 600;

const ENV_DB_URL: &str = "TRUSTD_DB_URL";
const ENV_DB_NAME: &str = "TRUSTD_DB_NAME";
const ENV_DB_USER: &str = "TRUSTD_DB_USER";
const ENV_DB_PASS: &str = "TRUSTD_DB_PASSWORD";
const ENV_DB_HOST: &str = "TRUSTD_DB_HOST";
const ENV_DB_PORT: &str = "TRUSTD_DB_PORT";
const ENV_DB_MAX_CONN: &str = "TRUSTD_DB_MAX_CONN";
const ENV_DB_MIN_CONN: &str = "TRUSTD_DB_MIN_CONN";
const ENV_DB_CONNECT_TIMEOUT: &str = "TRUSTD_DB_MIN_CONN";
const ENV_DB_ACQUIRE_TIMEOUT: &str = "TRUSTD_DB_ACQUIRE_TIMEOUT";
const ENV_DB_MAX_LIFETIME: &str = "TRUSTD_DB_MAX_LIFETIME";
const ENV_DB_IDLE_TIMEOUT: &str = "TRUSTD_DB_IDLE_TIMEOUT";
const ENV_DB_SSLMODE: &str = "TRUSTD_DB_SSLMODE";

/// PostgreSQL SSL mode
#[derive(Copy, Clone, Debug, Default, clap::ValueEnum, Eq, PartialEq, strum::Display)]
#[strum(serialize_all = "kebab-case")]
pub enum SslMode {
    Disable,
    Allow,
    #[default]
    Prefer,
    Require,
    VerifyCa,
    VerifyFull,
}

/// Database options
#[derive(clap::Parser, Debug, Clone, Eq, PartialEq)]
#[command(next_help_heading = "Database")]
#[group(id = "database")]
pub struct Database {
    /// A complete URL. Conflicts with the other database parameters.
    #[arg(id = "db-url", long, env = ENV_DB_URL)]
    pub url: Option<String>,
    #[arg(id = "db-user", long, env = ENV_DB_USER, default_value_t = DB_USER.into(), conflicts_with = "db-url")]
    pub username: String,
    #[arg(
        id = "db-password",
        long,
        env = ENV_DB_PASS,
        default_value_t = DB_PASS.into(),
    )]
    pub password: String,
    #[arg(id = "db-host", long, env = ENV_DB_HOST, default_value_t = DB_HOST.into(), conflicts_with = "db-url")]
    pub host: String,
    #[arg(id = "db-port", long, env = ENV_DB_PORT, default_value_t = DB_PORT.into(), conflicts_with = "db-url")]
    pub port: u16,
    #[arg(id = "db-name", long, env = ENV_DB_NAME, default_value_t = DB_NAME.into(), conflicts_with = "db-url")]
    pub name: String,
    #[arg(id = "db-max-conn", long, env = ENV_DB_MAX_CONN, default_value_t = DB_MAX_CONN.into(), conflicts_with = "db-url")]
    pub max_conn: u32,
    #[arg(id = "db-min-conn", long, env = ENV_DB_MIN_CONN, default_value_t = DB_MIN_CONN.into(), conflicts_with = "db-url")]
    pub min_conn: u32,
    #[arg(id="db-sslmode", long, env = ENV_DB_SSLMODE, default_value_t, conflicts_with = "db-url", value_enum)]
    pub sslmode: SslMode,

    #[arg(id="db-conn-timeout", long, env = ENV_DB_CONNECT_TIMEOUT, default_value_t=DB_CONNECT_TIMEOUT.into(), conflicts_with = "db-url")]
    pub connect_timeout: u64,
    #[arg(id="db-acquire-timeout", long, env = ENV_DB_ACQUIRE_TIMEOUT, default_value_t=DB_ACQUIRE_TIMEOUT.into(), conflicts_with = "db-url")]
    pub acquire_timeout: u64,
    #[arg(id="db-max-lifetime", long, env = ENV_DB_MAX_LIFETIME, default_value_t=DB_MAX_LIFETIME.into(), conflicts_with = "db-url")]
    pub max_lifetime: u64,
    #[arg(id="db-idle-timeout", long, env = ENV_DB_IDLE_TIMEOUT, default_value_t=DB_IDLE_TIMEOUT.into(), conflicts_with = "db-url")]
    pub idle_timeout: u64,
}

impl Database {
    pub fn from_env() -> Result<Database, anyhow::Error> {
        Ok(Database {
            url: env::var(ENV_DB_URL).ok(),
            username: env::var(ENV_DB_USER).unwrap_or(DB_USER.into()),
            password: env::var(ENV_DB_PASS).unwrap_or(DB_PASS.into()),
            name: env::var(ENV_DB_NAME).unwrap_or(DB_NAME.into()),
            host: env::var(ENV_DB_HOST).unwrap_or(DB_HOST.into()),
            port: match env::var(ENV_DB_PORT) {
                Ok(s) => s.parse::<u16>()?,
                _ => DB_PORT,
            },
            max_conn: match env::var(ENV_DB_MAX_CONN) {
                Ok(s) => s.parse::<u32>()?,
                _ => DB_MAX_CONN,
            },
            min_conn: match env::var(ENV_DB_SSLMODE) {
                Ok(s) => s.parse::<u32>()?,
                _ => DB_MIN_CONN,
            },
            connect_timeout: match env::var(ENV_DB_CONNECT_TIMEOUT) {
                Ok(s) => parse_duration(&s)
                    .unwrap_or(DB_IDLE_TIMEOUT.std_seconds())
                    .as_secs(),
                _ => DB_CONNECT_TIMEOUT,
            },
            acquire_timeout: match env::var(ENV_DB_ACQUIRE_TIMEOUT) {
                Ok(s) => parse_duration(&s)
                    .unwrap_or(DB_ACQUIRE_TIMEOUT.std_seconds())
                    .as_secs(),
                _ => DB_ACQUIRE_TIMEOUT,
            },
            max_lifetime: match env::var(ENV_DB_MAX_LIFETIME) {
                Ok(s) => parse_duration(&s)
                    .unwrap_or(DB_MAX_LIFETIME.std_seconds())
                    .as_secs(),
                _ => DB_MAX_LIFETIME,
            },
            idle_timeout: match env::var(ENV_DB_IDLE_TIMEOUT) {
                Ok(s) => parse_duration(&s)
                    .unwrap_or(DB_IDLE_TIMEOUT.std_seconds())
                    .as_secs(),
                _ => DB_IDLE_TIMEOUT,
            },
            sslmode: match env::var(ENV_DB_SSLMODE) {
                Ok(s) => SslMode::from_str(&s, false)
                    .map_err(|s| anyhow!("Failed to convert '{s}' to SslMode"))?,
                _ => Default::default(),
            },
        })
    }

    pub fn to_url(&self) -> String {
        if let Some(url) = &self.url {
            return url.clone();
        }

        format!(
            "postgres://{username}:{password}@{host}:{port}/{db_name}?sslmode={sslmode}",
            username = &self.username,
            password = &self.password,
            host = &self.host,
            port = self.port,
            db_name = &self.name,
            sslmode = &self.sslmode,
        )
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use clap::Parser;

    #[test]
    fn url() {
        let result = Database::try_parse_from(["test", "--db-url", "postgres://localhost:4321"])
            .expect("must parse");

        assert_eq!(
            Database {
                url: Some("postgres://localhost:4321".to_string()),
                username: DB_USER.into(),
                password: DB_PASS.into(),
                host: DB_HOST.into(),
                port: DB_PORT,
                name: DB_NAME.into(),
                max_conn: DB_MAX_CONN,
                min_conn: DB_MIN_CONN,
                connect_timeout: DB_CONNECT_TIMEOUT,
                acquire_timeout: DB_ACQUIRE_TIMEOUT,
                max_lifetime: DB_MAX_LIFETIME,
                idle_timeout: DB_IDLE_TIMEOUT,
                sslmode: SslMode::default(),
            },
            result
        );
    }

    #[test]
    fn args() {
        let result =
            Database::try_parse_from(["test", "--db-sslmode", "disable"]).expect("must parse");

        assert_eq!(
            Database {
                url: None,
                username: DB_USER.into(),
                password: DB_PASS.into(),
                host: DB_HOST.into(),
                port: DB_PORT,
                name: DB_NAME.into(),
                max_conn: DB_MAX_CONN,
                min_conn: DB_MIN_CONN,
                connect_timeout: DB_CONNECT_TIMEOUT,
                acquire_timeout: DB_ACQUIRE_TIMEOUT,
                max_lifetime: DB_MAX_LIFETIME,
                idle_timeout: DB_IDLE_TIMEOUT,
                sslmode: SslMode::Disable,
            },
            result
        );

        assert_eq!(
            result.to_url(),
            "postgres://postgres:trustify@localhost:5432/trustify?sslmode=disable"
        );
    }
}
